flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/MatchCodeGenerator.scala

```diff
@@ -0,0 +1,320 @@
+import org.apache.flink.table.codegen.Indenter.toISC
+import org.apache.flink.table.plan.schema.RowSchema
+import org.apache.flink.util.Collector
+import org.apache.flink.util.MathUtils.checkedDownCast
+import java.lang.{Long => JLong}
+
+import scala.collection.JavaConverters._
+
+/**
+  * A code generator for generating CEP related functions.
+  *
+  * @param config configuration that determines runtime behavior
+  * @param nullableInput input(s) can be null.
+  * @param input type information about the first input of the Function
+  * @param currentPattern if generating condition the name of pattern, which the condition will
+  *                       be applied to
+  */
+class MatchCodeGenerator(
+    config: TableConfig,
+    nullableInput: Boolean,
+    input: TypeInformation[_ <: Any],
+    currentPattern: Option[String] = None)
+  extends CodeGenerator(config, nullableInput, input){
+
+  def generateMatchFunction[F <: Function, T <: Any](
+    name: String,
+    clazz: Class[F],
+    bodyCode: String,
+    returnType: TypeInformation[T])
+  : GeneratedFunction[F, T] = {
+    // This is a separate method from FunctionCodeGenerator#generateFunction because as of now
+    // functions in CEP library do not support rich interfaces
+    val funcName = newName(name)
+    val collectorTypeTerm = classOf[Collector[Any]].getCanonicalName
+    val (functionClass, signature, inputStatements, isInterface) =
+      if (clazz == classOf[IterativeCondition[_]]) {
+        val baseClass = classOf[IterativeCondition[_]]
+        val inputTypeTerm = boxedTypeTermForTypeInfo(input)
+        val contextType = classOf[IterativeCondition.Context[_]].getCanonicalName
+
+        (baseClass,
+          s"boolean filter( Object _in1, $contextType $contextTerm)",
+          List(s"$inputTypeTerm $input1Term = ($inputTypeTerm) _in1;"),
+          false)
+      } else if (clazz == classOf[PatternSelectFunction[_, _]]) {
+        val baseClass = classOf[PatternSelectFunction[_, _]]
+        val inputTypeTerm =
+          s"java.util.Map<String, java.util.List<${boxedTypeTermForTypeInfo(input)}>>"
+
+        (baseClass,
+          s"Object select($inputTypeTerm $input1Term)",
+          List(),
+          true)
+      } else if (clazz == classOf[PatternFlatSelectFunction[_, _]]) {
+        val baseClass = classOf[PatternFlatSelectFunction[_, _]]
+        val inputTypeTerm =
+          s"java.util.Map<String, java.util.List<${boxedTypeTermForTypeInfo(input)}>>"
+
+        (baseClass,
+          s"void flatSelect($inputTypeTerm $input1Term, $collectorTypeTerm $collectorTerm)",
+          List(),
+          true)
+      } else {
+        throw new CodeGenException("Unsupported Function.")
+      }
+
+    val extendsKeyword = if (isInterface) "implements" else "extends"
+    val funcCode = j"""
+      |public class $funcName $extendsKeyword ${functionClass.getCanonicalName} {
+      |
+      |  ${reuseMemberCode()}
+      |
+      |  public $funcName() throws Exception {
```

Comment:
Adding support of `open/close` should be done carefully in CEP library because of:
1. backwards compatibility (we previously serialized conditions) - changes to those interfaces have to be done carefully
2. we want to be able to add dynamic patterns in the future, so at some point we have to open also on control message arrival

Therefore my suggestion for the BETA release of MATCH_RECOGNIZE we should forbid using any "Rich" interfaces.